### PR TITLE
Fix ACInterface residual and on-diagonal Jacobian

### PR DESCRIPTION
**Pull request by dschwen**
This addresses part of the problem in #5758. Another PR will follow to finish the work on the off-diagonal Jacobian and the support for variable dependent kappas.

### CODE REVIEW

**Comment by permcody** [comment]
@laagesen - Could you review this PR before we merge it? Thanks

**Comment by laagesen** [comment]
Sure I will have a look.

File: modules/phase_field/src/kernels/ACInterface.C [hunk 1]
```diff
@@ -9,29 +9,60 @@
...
+    _gradarg(_nvar)
 {
-  // Get number of coupled variables
-  unsigned int nvar = _coupled_moose_vars.size();
+  // Get mobility and kappa derivatives and coupled variable gradients
+  for (unsigned int i = 0; i < _nvar; ++i)
+  {
+    MooseVariable *ivar = _coupled_moose_vars[i];
+
+    _dLdarg[i] = &getMaterialPropertyDerivative<Real>("mob_name", ivar->name());
```
Comment (laagesen) [inline modules/phase_field/src/kernels/ACInterface.C]: Can we use `validateNonlinearCoupling` before this to warn the user that they need to specify any required `args` in the input file for these derivatives?

File: modules/phase_field/src/kernels/ACInterface.C [hunk 2]
```diff
@@ -41,17 +72,70 @@ ACInterface::initialSetup()
...
+  RealGradient g = _grad_u[_qp] * _dkappadop[_qp];
+  for (unsigned int i = 0; i < _nvar; ++i)
+    g += (*_gradarg[i])[_qp] * (*_dkappadarg[i])[_qp];
+  return g;
+}
+
+RealGradient
+ACInterface::nablaLKappaPsi()
+{
+  // sum is the product rule gradient \f$ \nable (L\kappa\psi) \f$
```
Comment (laagesen) [inline modules/phase_field/src/kernels/ACInterface.C]: typo `\nable` -> `\nabla`

File: modules/phase_field/src/kernels/ACInterface.C (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/src/kernels/ACInterface.C]: No. The validation needs to be in ```ACInterface::initialSetup()```, where it already is! **But** we need to make the ```validateCoupling``` to make sure that we _also_ validate that all AuxVariables are coupled (they could affect the gradient of L or kappa!).

I will make that change.

**Comment by laagesen** [comment]
Everything else looks good--the only other question I had was, is there anything that needs to be done to modify boundary conditions to account for the fact that L is now variable? The form of the boundary term that results from applying the product rule/divergence theorem doesn't change due to the fact that L can vary, so I suspect it shouldn't be a problem, but just thought I would ask.

**Comment by jwpeterson** [comment]
> The form of the boundary term that results from applying the product rule/divergence theorem doesn't change due to the fact that L can vary, so I suspect it shouldn't be a problem, but just thought I would ask.

The form of the term doesn't change, but it might need additional `computeQpJacobian()` terms added.

**Comment by dschwen** [comment]
I see. We do not have such a boundary condition yet. After talking with @laagesen we think that the existing boundary conditions (such as NeumannBC) are not affected by this as they have no notion of a _Mobility_. If we wanted to impose a physical _flux_ boundary condition we'd set the value of grad(c)*M _not_ just the value of grad(c). And this would apply to a Cahn-Hilliard problem rather than to an AC problem. I will open an issue to track this future development.

**Pull request merged**